### PR TITLE
Add option to set language in admin user view

### PR DESCRIPTION
**Pull request by 6543**
image
`/admin/users/<UserID>/edit`

image
`/admin/users/<UserID>`

---
*Sponsored by Kithara Software GmbH*

### CODE REVIEW

**Comment by lunny** [comment]
I don't think the title is related with the content.

**Comment by 6543** [comment]
uh sory i guess i git the wrong title :D

**Comment by lunny** [comment]
Why users don't change the language themselves but need admin to help? I can't imagine the scenes

**Comment by delvh** [comment]
Yeah, same as @lunny.
The usecase isn't clear to me either.
Instead, it sounds like something that can be abused when some admin suddenly switches your Gitea language to arabic as a prank

**Comment by 6543** [comment]
@lunny @delvh if an admin does such a "joke" you should not consider to use that instance at all !!!

and yes if you are not a self-registration instance but you get an account created it's possible that you have to set the language.

for now this is done by log in as the user after it has been created and then change the language.
but you loose the feature to let the user be forced to have to set a new password.

**Comment by 6543** [comment]
and not all people do speak englisch tbh ...

File: services/forms/admin.go [hunk 1]
```diff
@@ -41,6 +41,7 @@ type AdminEditUserForm struct {
 	Password                string `binding:"MaxSize(255)"`
 	Website                 string `binding:"ValidUrl;MaxSize(255)"`
 	Location                string `binding:"MaxSize(50)"`
+	Language                string `binding:"MaxSize(5)"`
```
Comment (delvh) [inline services/forms/admin.go:44]: MaxSize `5`?

Comment (delvh) [reply services/forms/admin.go:44]:
```suggestion
	Language                string `binding:"MaxSize(10)"`
```
even if we don't use it, give it a bit of space for the eventuality.

File: templates/admin/user/edit.tmpl [hunk 2]
```diff
@@ -70,6 +70,21 @@
...
+					<div class="ui selection dropdown">
+						<input name="language" type="hidden" value="{{.User.Language}}">
+						{{svg "octicon-triangle-down" 14 "dropdown icon"}}
+						<div class="text">{{range .AllLangs}}{{if eq $.User.Language .Lang}}{{.Name}}{{end}}{{end}}</div>
+						<div class="menu">
+						{{range .AllLangs}}
+							<div class="item{{if eq $.User.Language .Lang}} active selected{{end}}" data-value="{{.Lang}}">{{.Name}}</div>
+						{{end}}
+						</div>
+					</div>
```
Comment (delvh) [inline templates/admin/user/edit.tmpl:85]: Can't we extract a common template with the site footer?

File: services/forms/admin.go (hunk 1 above)
Comment (6543) [reply services/forms/admin.go:44]: we use the short identifier de-DE in the backend

File: templates/admin/user/edit.tmpl (hunk 2 above)
Comment (6543) [reply templates/admin/user/edit.tmpl:85]: its similar but differ

File: services/forms/admin.go (hunk 1 above)
Comment (6543) [reply services/forms/admin.go:44]: ->
https://github.com/go-gitea/gitea/blob/e7cb8da2a8310ac167b6f613b283caa3316a7154/models/user/user.go#L94

File: templates/admin/user/edit.tmpl (hunk 2 above)
Comment (6543) [reply templates/admin/user/edit.tmpl:85]:
```diff
diff --git a/user_settings_appearance.tmpl b/admin_user_edit.tmpl
index 3995b58..3a421d1 100644
--- a/user_settings_appearance.tmpl
+++ b/admin_user_edit.tmpl
@@ -1,12 +1,8 @@
-<div class="field">
-       <div class="ui language selection dropdown" id="language">
-               <input name="language" type="hidden" value="{{.SignedUser.Language}}">
-               {{svg "octicon-triangle-down" 14 "dropdown icon"}}
-               <div class="text">{{range .AllLangs}}{{if eq $.SignedUser.Language .Lang}}{{.Name}}{{end}}{{end}}</div>
-               <div class="menu">
-                       {{range .AllLangs}}
-                               <div class="item{{if eq $.SignedUser.Language .Lang}} active selected{{end}}" data-value="{{.Lang}}">{{.Name}}</div>
-                       {{end}}
-               </div>
+{{if .User.Language}}
+       <div class="flex-item-body">
+               <span class="flex-text-inline">
+                       <b>{{ctx.Locale.Tr "settings.language"}}:</b>
+                       {{range .AllLangs}}{{if eq $.User.Language .Lang}}{{.Name}}{{end}}{{end}}
+               </span>
        </div>
-</div>
+{{end}}
```
```diff
diff --git a/admin_user_view_details.tmpl b/user_settings_appearance.tmpl
index 3a421d1..3995b58 100644
--- a/admin_user_view_details.tmpl
+++ b/user_settings_appearance.tmpl
@@ -1,8 +1,12 @@
-{{if .User.Language}}
-       <div class="flex-item-body">
-               <span class="flex-text-inline">
-                       <b>{{ctx.Locale.Tr "settings.language"}}:</b>
-                       {{range .AllLangs}}{{if eq $.User.Language .Lang}}{{.Name}}{{end}}{{end}}
-               </span>
+<div class="field">
+       <div class="ui language selection dropdown" id="language">
+               <input name="language" type="hidden" value="{{.SignedUser.Language}}">
+               {{svg "octicon-triangle-down" 14 "dropdown icon"}}
+               <div class="text">{{range .AllLangs}}{{if eq $.SignedUser.Language .Lang}}{{.Name}}{{end}}{{end}}</div>
+               <div class="menu">
+                       {{range .AllLangs}}
+                               <div class="item{{if eq $.SignedUser.Language .Lang}} active selected{{end}}" data-value="{{.Lang}}">{{.Name}}</div>
+                       {{end}}
+               </div>
        </div>
-{{end}}
+</div>
```
i don't see a common template here ?!?

**Review by delvh — APPROVED** [review]

File: templates/admin/user/edit.tmpl [hunk 3]
```diff
@@ -70,6 +70,21 @@
 					<input id="password" name="password" type="password" autocomplete="new-password">
 					<p class="help">{{ctx.Locale.Tr "admin.users.password_helper"}}</p>
 				</div>
+
+				<div class="field {{if .Err_Language}}error{{end}}">
+					<label for="language">{{ctx.Locale.Tr "settings.language"}}</label>
+					<div class="ui selection dropdown">
+						<input name="language" type="hidden" value="{{.User.Language}}">
+						{{svg "octicon-triangle-down" 14 "dropdown icon"}}
+						<div class="text">{{range .AllLangs}}{{if eq $.User.Language .Lang}}{{.Name}}{{end}}{{end}}</div>
```
Comment (KN4CK3R) [inline templates/admin/user/edit.tmpl:79]: This should not be needed. The dropdown should get the selected value from the hidden input field.

Comment (6543) [reply templates/admin/user/edit.tmpl:79]: @KN4CK3R no you need it - I tested it and if i remove that line the closed drop down has no value

**Review by denyskon — APPROVED** [review]

**Comment by delvh** [comment]
@6543 please update the branch

**Pull request merged**